lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+            "session_function.jinja"
+        )
+        session_function = SESSION_FUNCTION_TEMPLATE.render(
+            indentation_block=indentation_block,
+            session_function_name=session_function_name,
+            session_function_body=session_function_body,
+            return_dict_name=return_dict_name,
+        )
+
+        # Generate calculation code block for the session
+        # This is to be used in multi-session module
+        session_calculation = (
+            f"{indentation_block}artifacts.update({session_function_name}())"
+        )
+
+        return {
+            "session_imports": session_imports,
+            "artifact_functions": artifact_functions,
+            "session_function": session_function,
+            "session_function_return": session_function_return,
+            "session_calculation": session_calculation,
+        }
+
+    def _compose_module(
+        self,
+        session_artifacts_sorted: List[SessionArtifacts],
+        indentation: int = 4,
+    ) -> str:
+        """
+        Generate a Python module that calculates artifacts
+        in the given artifact collection.
+        """
+        indentation_block = " " * indentation
+
+        # Initiate store for module script components
+        module_dict: Dict[str, List[str]] = {
+            "session_imports": [],
+            "artifact_functions": [],
+            "session_function": [],
+            "session_function_return": [],
+            "session_calculation": [],
+        }
+
+        # Extract module script components by session
+        for session_artifacts in session_artifacts_sorted:
+            session_module_dict = self._extract_session_module(
+                session_artifacts=session_artifacts,
+                indentation=indentation,
+            )
+            module_dict["session_imports"].append(
+                session_module_dict["session_imports"]
+            )
+            module_dict["artifact_functions"].append(
+                session_module_dict["artifact_functions"]
+            )
+            module_dict["session_function"].append(
+                session_module_dict["session_function"]
+            )
+            module_dict["session_function_return"].append(
+                session_module_dict["session_function_return"]
+            )
+            module_dict["session_calculation"].append(
+                session_module_dict["session_calculation"]
+            )
+
+        # Combine components by "type"
+        module_imports = "\n".join(module_dict["session_imports"])
+        artifact_functions = "\n\n".join(module_dict["artifact_functions"])
+        session_functions = "\n".join(module_dict["session_function"])
+        module_function_body = "\n".join(module_dict["session_calculation"])
+        module_function_return = ", ".join(
+            module_dict["session_function_return"]
+        )
+
+        # Put all together to generate module text
+        # TODO: Replace with jinja template
```

Comment:
Look like this TODO is done? :)

Reply:
Yeah, seems like so (thanks to MJL); will remove the comment